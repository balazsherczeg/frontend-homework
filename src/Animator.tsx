import React, { useState, useEffect } from "react";

// https://stackoverflow.com/a/54114180/9162948

function useDelayUnmount(isMounted: boolean, delayTime: number) {
  const [shouldRender, setShouldRender] = useState(false);

  useEffect(() => {
    let timeoutId: number;
    if (isMounted && !shouldRender) {
      setShouldRender(true);
    } else if (!isMounted && shouldRender) {
      timeoutId = window.setTimeout(() => setShouldRender(false), delayTime);
    }
    return () => clearTimeout(timeoutId);
  }, [isMounted, delayTime, shouldRender]);

  return shouldRender;
}

type Props = {
  children: React.ReactNode,
  className: string,
  delay?: number,
  isMounted: boolean,
}

const Animator = ({
  children,
  className,
  delay = 250,
  isMounted,
}: Props) => {
  const shouldRenderChild = useDelayUnmount(isMounted, delay);
  const animationClassName = isMounted ? `${className}--mounted` : `${className}--unmounted`

  return (
    <div
      className={`${className} ${animationClassName}`}
    >
      {shouldRenderChild && children}
    </div>
  );
};

export default Animator;
