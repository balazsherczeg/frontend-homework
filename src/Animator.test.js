import { mount } from 'enzyme'
import Animator from './Animator'

describe('<Animator />', () => {
  const render = (props) => mount( // Any better way to test component with a useEffect hook?
    <Animator className="Test" {...props}><div>a child</div></Animator>
  )

  it('renders its child if mounted', () => {
    expect(render({ isMounted: true })).toMatchSnapshot()
  })

  it('doesnâ€™t render its child if unmounted', () => {
    expect(render({ isMounted: false })).toMatchSnapshot()
  })
})
