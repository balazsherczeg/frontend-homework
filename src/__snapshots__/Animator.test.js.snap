// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Animator /> doesnâ€™t render its child if unmounted 1`] = `
<Animator
  className="Test"
  isMounted={false}
>
  <div
    className="Test Test--unmounted"
  />
</Animator>
`;

exports[`<Animator /> renders its child if mounted 1`] = `
<Animator
  className="Test"
  isMounted={true}
>
  <div
    className="Test Test--mounted"
  >
    <div>
      a child
    </div>
  </div>
</Animator>
`;
